scripts/llext_link_helper.py

```diff
@@ -32,56 +33,107 @@ def parse_args():
 def main():
 	parse_args()
 
+	TYPE_NONE = 0
+	TYPE_DATA = 1
+	TYPE_BSS = 2
+	last_type = TYPE_NONE
+
+	SHF_ALLOC = 2
+	SHF_EXECINSTR = 4
+
 	elf = ELFFile(open(args.file, 'rb'))
 
 	text_addr = int(args.text_addr, 0)
-	p = re.compile(r'(^lib|\.so$)')
-	module = p.sub('', args.file)
 
-	if elf.num_segments() != 0:
-		# A shared object type image, it contains segments
-		sections = ['.text', '.rodata', '.data', '.bss']
-		alignment = [0x1000, 0x1000, 0x10, 0x1000]
+	# File names differ when building shared or relocatable objects
+	if args.file[:-3] == '.so':
+		p = re.compile(r'(^lib|\.so$)')
+		fname = args.file
 	else:
-		# A relocatable object, need to handle all sections separately
-		sections = ['.text',
-			    f'._log_const.static.log_const_{module}_',
-			    '.static_uuids', '.z_init_APPLICATION90_0_', '.module',
-			    '.mod_buildinfo', '.data', '.trace_ctx', '.bss']
-		alignment = [0x1000, 0x1000, 0x0, 0x0, 0x0, 0x0, 0x10, 0x0, 0x1000]
+		fpath = pathlib.Path(args.file)
+		fname = fpath.name
+		p = re.compile(r'(^lib|_llext_lib\.obj$)')
+	module = p.sub('', fname)
 
 	last_increment = 0
 
-	command = [args.command]
+	command = [args.command, f'-Wl,-Ttext=0x{text_addr:x}']
+	data_found = False
+	bss_found = False
+
+	# The layout seems to be: .text, followed by other .data, .rodata and
+	# similar sections with the SHF_ALLOC flag set, and finally .bss. Then
+	# come all sections with no SHF_ALLOC set like .comment and similar
+	for section in elf.iter_sections():
+		s_flags = section.header['sh_flags']
+		s_type = section.header['sh_type']
+		s_name = section.name
+		s_size = section.header['sh_size']
+		s_offset = section.header['sh_offset']
+
+		if (s_flags & (SHF_ALLOC | SHF_EXECINSTR) == SHF_ALLOC | SHF_EXECINSTR and
+		    s_type == 'SHT_PROGBITS'):
+			if s_name != '.text':
+				print(f"Warning! Non-standard executable section {s_name}")
+
+			if bss_found or data_found:
+				print("ERROR: So far .text is expected first. Fix the script!")
```

Comment:
which script? This one?

Reply:
@marc-hb yes

Reply:
Then the usual "FIXME"

```suggestion
			assert not bss_found and not data_found, "FIXME: So far .text is expected first"
```